### DISCUSSION TITLE
openssl (3.x) can not connect to the "test.openquantumsafe.org" and create server with "Openssl s_server" with OQS signature

### DISCUSSION THREAD

**Discussion by k-Artin** [open-quantum-safe/oqs-provider#102] [Q&A]
Hello
I want to connect to the https://test.openquantumsafe.org/ with OQS signature for example with this command:

echo "GET /" | openssl s_client -connect test.openquantumsafe.org:6145 -groups kyber512 -CAfile CA.crt

i got this error:

SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl/record/rec_layer_s3.c:1605:SSL alert number 40

and when i want to create server with "OpenSSL s_server" for example with this command:

openssl s_server -port 2233 -tls1_3 -key dilithium2_srv.key -cert dilithium2_srv_rsa.crt -CAfile ca_rsa.crt -curves kyber512 -ciphersuites TLS_AES_256_GCM_SHA384 -provider-path /home/oqs-provider/_build/oqsprov/ -provider oqsprovider -provider default 

i got this error:

SSL routines:ssl_set_cert:unknown certificate type:ssl/ssl_rsa.c:239

the certificate signing with rsa key and the server key is dilithium2.

openssl version is OpenSSL 3.1.0-beta1 and oqs provider version is  0.5.0-dev.

I think I forgot something in installing it or using openssl s_server.
Thank you in advance!

**Comment by baentsch** [comment]
Nope, you didn't forget anything. You stumbled upon https://github.com/open-quantum-safe/oqs-provider/issues/9 / https://github.com/openssl/openssl/issues/10512 .

You could validate that if you can confirm successful connection to port 6007 (bypassing the (provider-based) TLS signature logic).

Reminds me to update oqs-provider to work with https://github.com/openssl/openssl/pull/19312.

**Reply by k-Artin** [reply]
i can connect with all classic signatures like rsa and ecdsa. but when using the OQS signature i got the error SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl/record/rec_layer_s3.c:1605:SSL alert number 40.
i also using the openquantumsafe/oqs-ossl3:ietf115 (docker image) but stil i got all error above.
Thank you in advance!

**Reply by baentsch** [reply]
Please try again after pulling the new `openquantumsafe/oqs-ossl3:ietf115` image.

**Comment by baentsch** [comment]
> i also using the openquantumsafe/oqs-ossl3:ietf115 (docker image) but stil i got all error above.

That is still expected as OpenSSL master did not yet merge https://github.com/openssl/openssl/pull/19312. 

But I like your thinking: Give me a bit of time and I'll create a branch in oqs-provider that works with 19312 and put that into the ietf115 image.

**Reply by baentsch** [reply]
@k-Artin : oqs-provider branch working with https://github.com/openssl/openssl/pull/19312  now available: https://github.com/open-quantum-safe/oqs-provider/tree/ossl-sigload and IETF115 docker image lifted to that code level:

`docker run -it openquantumsafe/oqs-ossl3:ietf115 bash -c 'apt install -y wget && wget https://test.openquantumsafe.org/CA.crt && echo "GET /" | openssl s_client -connect test.openquantumsafe.org:6145 -groups kyber512 -CAfile CA.crt'` resulting in:

```
[...]
SSL handshake has read 5545 bytes and written 1218 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 128 bit
This TLS version forbids renegotiation.
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
DONE

```

**Reply by k-Artin** [reply]
thanks the docker image is working but i have some question about that.
when i am building oqs-provider with openssl 3.1.0-beta i got the same error.
this changing only work with openssl master (3.2.0-dev) ??

**Reply by baentsch** [reply]
> this changing only work with openssl master (3.2.0-dev) ??

No. The answer is stated above:

> That is still expected as OpenSSL master did not yet merge https://github.com/openssl/openssl/pull/19312.

Again, explicitly: Until https://github.com/openssl/openssl/issues/10512 is closed (by merging https://github.com/openssl/openssl/pull/19312) OpenSSL cannot handle PQ sigs.
